micro_http/src/connection.rs

```diff
@@ -0,0 +1,714 @@
+    }
+
+    fn shift_buffer_left(&mut self, line_start_index: usize, end_cursor: usize) {
+        // We don't want to shift something that is already at the beginning.
+        if line_start_index != 0 {
+            // Move the bytes from `line_start_index` to the beginning of the buffer.
+            for cursor in 0..(end_cursor - line_start_index) {
+                self.buffer[cursor] = self.buffer[line_start_index + cursor];
+            }
+
+            // Clear the rest of the buffer.
+            for cursor in (end_cursor - line_start_index)..end_cursor {
+                self.buffer[cursor] = 0;
+            }
+        }
+
+        // Update `read_cursor`.
+        self.read_cursor = end_cursor - line_start_index;
+    }
+
+    /// Returns the first parsed request in the queue.
+    pub fn pop_parsed_request(&mut self) -> Option<Request> {
+        self.parsed_requests.pop_front()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::Version;
+    use super::*;
+    use std::os::unix::net::UnixStream;
+
+    #[test]
+    #[allow(clippy::cognitive_complexity)]
+    fn test_try_read_connection() {
+        // Test request with `Expect` header.
+        if let Ok((mut sender, receiver)) = UnixStream::pair() {
+            receiver.set_nonblocking(true).expect("Can't modify socket");
+            let mut conn = HttpConnection::new(receiver);
+            sender
+                .write_all(
+                    b"PATCH http://localhost/home HTTP/1.1\r\n \
+                                     Expect: 100-continue\r\n \
+                                     Content-Length: 26\r\n \
+                                     Transfer-Encoding: chunked\r\n\r\n",
+                )
+                .unwrap();
+            assert!(conn.try_read().is_ok());
+
+            sender.write_all(b"this is not\n\r\na json \nbody").unwrap();
+            conn.try_read().unwrap();
+            let request = conn.pop_parsed_request().unwrap();
+            let expected_request = Request::try_from(
+                b"PATCH http://localhost/home HTTP/1.1\r\n \
+                                     Expect: 100-continue\r\n \
+                                     Transfer-Encoding: chunked\r\n \
+                                     Content-Length: 26\r\n\r\nthis is not\n\r\na json \nbody",
+            )
+            .unwrap();
+            assert_eq!(request, expected_request);
+        }
+
+        // Long request headers.
+        if let Ok((mut sender, receiver)) = UnixStream::pair() {
+            receiver.set_nonblocking(true).expect("Can't modify socket");
+            let mut conn = HttpConnection::new(receiver);
+            sender
+                .write_all(
+                    b"PATCH http://localhost/home HTTP/1.1\r\n \
+                                     Expect: 100-continue\r\n \
+                                     Transfer-Encoding: chunked\r\n",
+                )
+                .unwrap();
+
+            for _ in 0..90 {
```

Comment:
Having the same header more than once is actually not HTTP compliant. We should probably return error in this case. If you want to get long requests headers you can use dummy random names for the fields. 

From specification:
"Multiple message-header fields with the same field-name MAY be
   present in a message if and only if the entire field-value for that
   header field is defined as a comma-separated list [i.e., #(values)].
   It MUST be possible to combine the multiple header fields into one
   "field-name: field-value" pair, without changing the semantics of the
   message, by appending each subsequent field-value to the first, each
   separated by a comma. The order in which header fields with the same
   field-name are received is therefore significant to the
   interpretation of the combined field value, and thus a proxy MUST NOT
   change the order of these field values when a message is forwarded."